numpy/_core/src/umath/loops_trigonometric.dispatch.cpp

```diff
@@ -0,0 +1,298 @@
+                x, quadrant, codyw_pio2_highf, codyw_pio2_medf, codyw_pio2_lowf
+            );
+            vec_f32 reduced_x2 = hn::Mul(reduced_x, reduced_x);
+
+            // compute cosine and sine
+            vec_f32 cos = simd_cosine_poly_f32(reduced_x2);
+            vec_f32 sin = simd_sine_poly_f32(reduced_x, reduced_x2);
+
+            vec_s32 iquadrant = hn::NearestInt(quadrant);
+            if (trig_op == SIMD_COMPUTE_COS) {
+                iquadrant = hn::Add(iquadrant, ones);
+            }
+            // blend sin and cos based on the quadrant
+            opmask_t sine_mask = hn::RebindMask(f32, hn::Eq(hn::And(iquadrant, ones), hn::Zero(s32)));
+            cos = hn::IfThenElse(sine_mask, sin, cos);
+
+            // multiply by -1 for appropriate elements
+            opmask_t negate_mask = hn::RebindMask(f32, hn::Eq(hn::And(iquadrant, twos), twos));
+            cos = hn::MaskedSubOr(cos, negate_mask, zerosf, cos);
+            cos = hn::IfThenElse(nnan_mask, cos, hn::Set(f32, NPY_NANF));
+
+            if (sdst == 1) {
+                hn::StoreN(cos, f32, dst, len);
+            } else {
+                ScatterIndexN(cos, dst, sdst, len);
+            }
+        }
+        if (!hn::AllTrue(f32, simd_mask)) {
+            npy_uint64 simd_maski;
+            hn::StoreMaskBits(f32, simd_mask, (uint8_t*)&simd_maski);
+            float NPY_DECL_ALIGNED(NPY_SIMD_WIDTH) ip_fback[hn::Lanes(f32)];
+            hn::Store(x_in, f32, ip_fback);
+
+            // process elements using libc for large elements
+            if (trig_op == SIMD_COMPUTE_COS) {
+                for (unsigned i = 0; i < hn::Lanes(f32); ++i) {
+                    if ((simd_maski >> i) & 1) {
+                        continue;
+                    }
+                    dst[sdst*i] = npy_cosf(ip_fback[i]);
+                }
+            }
+            else {
+                for (unsigned i = 0; i < hn::Lanes(f32); ++i) {
+                    if ((simd_maski >> i) & 1) {
+                        continue;
+                    }
+                    dst[sdst*i] = npy_sinf(ip_fback[i]);
+                }
+            }
+        }
+    npyv_cleanup();
+    }
+}
+#endif // NPY_SIMD_FMA3
+
+/* Disable SIMD code sin/cos f64 and revert to libm: see
+ * https://mail.python.org/archives/list/numpy-discussion@python.org/thread/C6EYZZSR4EWGVKHAZXLE7IBILRMNVK7L/
+ * for detailed discussion on this*/
+#define DISPATCH_DOUBLE_FUNC(func) \
+NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(DOUBLE_##func) \
+(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(data)) \
+{ \
+    UNARY_LOOP { \
+        const npy_double in1 = *(npy_double *)ip1; \
+        *(npy_double *)op1 = npy_##func(in1); \
+    } \
+} \
+
+DISPATCH_DOUBLE_FUNC(sin)
+DISPATCH_DOUBLE_FUNC(cos)
+
+NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(FLOAT_sin)
+(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(data))
+{
+#if NPY_SIMD_F32 && NPY_SIMD_FMA3
```

Comment:
```suggestion
#if HWY_NATIVE_FMA
```

There's no `HWY_HAVE_FLOAT32`

Reply:
Do you know why we need NPY_SIMD_F32 ?